ARROW-5631: [C++] Fix FindBoost targets with cmake3.2

fsaintjacques opened the pull request:
cmake's FindBoost pre 3.5 did not export IMPORTED targets. This also fixes the newly introduced trusty docker image.

fsaintjacques commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -582,39 +582,43 @@ else()
+    set(ARROW_BOOST_LIBS ${BOOST_SYSTEM_LIBRARY} ${BOOST_FILESYSTEM_LIBRARY})
   endif()
+
+  if(ARROW_BOOST_USE_SHARED)
```
I'm not sure why this was like this, if someones wants to clarify, I'll gladly add a comment to the PR (@xhochy, @wesm).

kou replied:
This is needed to work with `BoostConfig.cmake` provided by Boost (not CMake).
The Boost package provided by MSYS2 installed `BoostConfig.cmake` at the time.
But the current Boost package doesn't install `BoostConfig.cmake`: https://github.com/msys2/MINGW-packages/pull/5326/files

See also: https://github.com/msys2/MINGW-packages/issues/5233

There is no platform that needs this. So we can remove this for now but we may need this again in the future.

fsaintjacques commented:
A note that this patch does not fully fix the cmake3.2 issue, see https://jira.apache.org/jira/browse/ARROW-5631?filter=-1 

But at least it gets a cmake build directory populated.

wesm commented:
Looks like @kou introduced that logic here

https://github.com/apache/arrow/commit/f014d76c68a9a977a19f87f7ea77511e4753bb8c

I'm not sure what it's for as we don't use this flag other places in our build system. Some projects use the flag to globally toggle between building SHARED/STATIC when using `add_library` but we are explicit about this everywhere.

fsaintjacques commented:
@kou afaik, find_package shouldn't be affected by BUILD_SHARED_LIBS?

wesm commented:
I think it's fine to keep the stashing; it seems like it was more likely future proofing

fsaintjacques commented:
Don't merge for now, it doesn't fix the docker cmake32 image.

fsaintjacques commented:
rebase with master, ready to land (once it passes tests). This passes locally, minus the linking error fixed by https://github.com/apache/arrow/pull/4635.

kou commented:
It seems that the current changes don't fix anything.
Should we really fix this problem or update title and description of this pull request?

fsaintjacques commented:
@kou previously cmake didn't even succeed at generating a build folder. The ctest failures are fixed in the PR linked in previous comment (#4635).

fsaintjacques commented:
@ursabot crossbox --help

fsaintjacques commented:
@ursabot crossbow --help

fsaintjacques commented:
@ursabot crossbow test

fsaintjacques commented:
@ursabot crossbow test -g docker

fsaintjacques commented:
@ursabot crossbow test docker docker-cpp docker-cpp-cmake32 docker-rust

fsaintjacques commented:
@ursabot crossbow test docker-cpp docker-cpp-cmake32 docker-rust

fsaintjacques commented:
@ursabot crossbow test docker-c_glib

fsaintjacques commented:
@kszucs do you know why the badge is at failed? The last one are waiting to be run.

fsaintjacques commented:
@kou could you check locally with this patch why c_glib docker image fails? I've exausted my debugging of  meson. I think that you need to namespace gandiva's `enums.c`.

```
gandiva-glib/meson.build:50:0: ERROR:  Tried to create target "enums.c", but a target of that name already exists.
```

I'll let you decide if renaming breaks anything.

fsaintjacques commented:
@ursabot crossbow test docker-c_glib

fsaintjacques commented:
@ursabot crossbow test docker-r

fsaintjacques commented:
@ursabot crossbow test -g docker

fsaintjacques commented:
This fixes the most pressing issues, notably
- docker-cpp
- docker-cmake32

I fixed also fixed docker-rust, and partly advanced on docker-r and docker-c_glib.

fsaintjacques commented:
@ursabot crossbow test docker-cpp docker-cpp-cmake32

kou commented on c_glib/Dockerfile:
```diff
@@ -40,7 +40,7 @@ ENV ARROW_BUILD_TESTS=OFF \
     ARROW_INSTALL_NAME_RPATH=OFF \
     LD_LIBRARY_PATH="${CONDA_PREFIX}/lib" \
-    PKG_CONFIG_PATH="${CONDA_PREFIX}/lib/pkgconfig" \
+    PKG_CONFIG_PATH="${CONDA_PREFIX}/lib/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig" \
```
Is this needed?
The added path is included in system `pkg-config` path. So it'll be searched without this.

wesm replied:
I'm going to revert this for now

kou commented:
BTW, I couldn't find any build log on CircleCI.

https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-46-circle-docker-c_glib shows

![image](https://user-images.githubusercontent.com/27350/59951456-f27f5a00-94b3-11e9-93e1-218a92ee7f1f.png)

And "build" link shows

![image](https://user-images.githubusercontent.com/27350/59951483-06c35700-94b4-11e9-88d7-10ba4c098670.png)

wesm commented on r/Dockerfile:
```diff
@@ -23,6 +23,8 @@ ENV CC=gcc \
 
 # r-base includes tzdata. Get around interactive stop in that package
 ENV DEBIAN_FRONTEND=noninteractive
+# workaround for install_github Github API rate limit
```
fixing
